test/test_helper.rb

```diff
@@ -110,6 +113,58 @@ def fake_machine(options = {})
   machine
 end
 
+class FakeHTTPProxyServer < WEBrick::HTTPProxyServer
+  def initialize(url, config)
+    @url = url
+    super(config)
+  end
+
+  def service(req, res)
+    req.attributes['url'] = @url
+    super(req, res)
+  end
+end
+
+module WEBrick
+  class BasicLog
+    def log(level, data)
+      return data.include? 'INFO  Basic Proxy Realm: user: authentication succeeded.'
+      # rubocop:disable GuardClause
+      # rubocop:disable UnreachableCode
+      if @log && level <= @level
+        data += "\n" if /\n\Z/ !~ data
+        @log << data
+      end
+    end
+  end
+end
+
+def fake_proxy_server(url, stream, options = {})
+  config = {
+    Port: options[:port],
+    Logger: WEBrick::Log.new(File.open(File::NULL, 'w')),
+    AccessLog: [[stream, '%m %U %s URL=%{url}n']]
+  }
+
+  if options[:authenticated]
+    htpasswd = WEBrick::HTTPAuth::Htpasswd.new options[:htpasswd_file]
+    htpasswd.set_passwd 'Proxy Realm', 'user', 'password'
+    htpasswd.flush
+    # Authenticator
+    authenticator = WEBrick::HTTPAuth::ProxyBasicAuth.new(
```

Comment:
Here is the cause for your logging issues. An Authenticator also takes a logger (see WEBrick::HTTPAuth::Authenticator) The config of the server is fine and goes to /dev/null resp. the StringIO instance, but since you don't pass a logger to the Authenticator stderr is used there.

Tip: If you run the test with the debugger and break in your custom patch for the log method, you get the full stack trace where this log message originates and it is straight forward from there.